File: ios/App/View/Games/Race/RaceEngine.swift
import Foundation

/// One traffic car. `x` is the center in field units, `y` the top edge;
/// `kind` picks the paint-job palette entry (part of the deterministic RNG
/// stream, like the asteroid silhouettes in Astro).
struct RaceTrafficCar: Equatable {
  /// Lane index 0 (left) to 2 (right).
  let lane: Int
  let x: Double
  var y: Double
  /// Own cruise speed (units/s); closing speed is player minus this.
  let speed: Double
  /// Paint-job palette index (0..trafficKindCount-1).
  let kind: Int
  /// Whether the overtake bonus for this car has been paid.
  var passed: Bool
}

/// Discrete things that happened during one `step(dt:)` — the native twin of
/// the web game's HUD/crash callbacks.
enum RaceEvent: Equatable {
  case overtake(total: Int)
  case crash(score: Int)
}

/// Pure port of the web RaceBot simulation
/// (`web/app/src/View/Games/Race/engine.ts`). No SwiftUI here: the engine is
/// a plain state machine driven by `step(dt:)` so it can be unit-tested
/// headlessly and rendered by any frontend.
///
/// Steering is tap-based (`steerLeft()`/`steerRight()` move the target lane);
/// nitro arrives as a held-input flag — the native twin of a held key.
///
/// Randomness (spawn lane, traffic speed, paint job) goes through an injected
/// closure so tests can pin every spawn decision. The RNG call order in
/// `trySpawnCar` (lane, speed, kind) matches the web engine exactly.
final class RaceEngine {
  // Field geometry and tuning — must stay byte-for-byte in sync with the web
  // constants in engine.ts.
  static let fieldWidth: Double = 420
  static let fieldHeight: Double = 720
  static let laneCount = 3
  /// Asphalt inset: shoulders on both sides of the three 110-unit lanes.
  static let roadLeft: Double = 45
  static let roadWidth: Double = 330
  static let laneWidth: Double = roadWidth / Double(laneCount)
  static let carWidth: Double = 62
  static let carLength: Double = 108
  /// Top edge of the player car; fixed — the world scrolls, the car does not.
  static let playerY: Double = 560
  /// Cruise speed at the green light (field units/s).
  static let baseSpeed: Double = 260
  /// Cruise ceiling; nitro can push past it.
  static let maxSpeed: Double = 640
  /// Cruise speed gained per second of driving — the difficulty ramp.
  static let acceleration: Double = 9
  /// Sideways glide speed during a lane change (units/s).
  static let laneChangeSpeed: Double = 520
  static let nitroMultiplier: Double = 1.6
  /// Gauge (0-1) drained per second while boosting.
  static let nitroDrain: Double = 0.35
  /// Gauge regenerated per second while cruising.
  static let nitroRegen: Double = 0.12
  /// Traffic cruises at a fixed speed rolled at spawn (units/s).
  static let trafficMinSpeed: Double = 110
  static let trafficMaxSpeed: Double = 215
  /// Distinct traffic paint jobs; the roll only picks the palette index.
  static let trafficKindCount = 4
  /// Seconds between spawn attempts at the start of a run…
  static let spawnIntervalStart: Double = 1.35
  /// …and the floor it ramps down to.
  static let spawnIntervalMin: Double = 0.55
  /// How much the spawn interval shrinks per second of driving.
  static let spawnIntervalRamp: Double = 0.008
  /// A spawn is skipped if it would leave no free lane in the top
  /// `entryWindow` units of the field — there must always be a way through.
  static let entryWindow: Double = 260
  /// Minimum bumper-to-bumper gap within a lane at spawn time.
  static let spawnGap: Double = carLength * 1.6
  /// Score per car passed.
  static let overtakeBonus = 50
  /// Field units per scored meter (≈ 43 m/s at base cruise).
  static let unitsPerMeter: Double = 6
  /// Collision paddings: slightly forgiving, like the web.
  static let hitWidth: Double = carWidth * 0.82
  static let hitLengthPad: Double = 6

  static let highScoreKey = "racebot-high-score"

  static func laneCenter(_ lane: Int) -> Double {
    roadLeft + laneWidth * (Double(lane) + 0.5)
  }

  /// Held input: nitro pedal down.
  var isBoosting = false

  private(set) var traffic: [RaceTrafficCar] = []
  /// Lane the player is gliding toward.
  private(set) var targetLane = 1
  /// Player center x; glides toward the target lane's center.
  private(set) var playerX = RaceEngine.laneCenter(1)
  /// Current cruise speed (before the nitro multiplier).
  private(set) var speed = RaceEngine.baseSpeed
  /// Nitro gauge, 0-1.
  private(set) var nitro: Double = 1
  private(set) var distanceMeters: Double = 0
  private(set) var overtakes = 0
  private(set) var isOver = false
  /// Engine clock in seconds; drives the difficulty ramp.
  private(set) var elapsed: Double = 0

  private var spawnCooldown = RaceEngine.spawnIntervalStart
  private let random: () -> Double

  init(random: @escaping () -> Double = { Double.random(in: 0..<1) }) {
    self.random = random
  }

  var score: Int {
    Int(distanceMeters) + overtakes * Self.overtakeBonus
  }

  /// True while the nitro flame should render and the multiplier applies.
  var isNitroActive: Bool {
    isBoosting && nitro > 0 && !isOver
  }

  /// Ground speed including nitro (what the speedometer shows).
  var effectiveSpeed: Double {
    speed * (isNitroActive ? Self.nitroMultiplier : 1)
  }

  /// Full reset (the web "Restart Race" path).
  func newGame() {
    traffic = []
    targetLane = 1
    playerX = Self.laneCenter(1)
    speed = Self.baseSpeed
    nitro = 1
    distanceMeters = 0
    overtakes = 0
    isOver = false
    elapsed = 0
    spawnCooldown = Self.spawnIntervalStart
    isBoosting = false
  }

  func steerLeft() {
    if !isOver, targetLane > 0 {
      targetLane -= 1
    }
  }

  func steerRight() {
    if !isOver, targetLane < Self.laneCount - 1 {
      targetLane += 1
    }
  }

  /// Advance the simulation. Mirrors the web `step` frame-for-frame: ramp
  /// the cruise speed, drain/refill nitro, glide the lane change, scroll
  /// traffic by the closing speed, pay overtakes, spawn on the cadence, and
  /// check for a crash. Returns the discrete events that occurred.
  func step(dt: Double) -> [RaceEvent] {
    guard !isOver else { return [] }
    var events: [RaceEvent] = []
    elapsed += dt

    // Cruise ramp and nitro gauge.
    speed = min(Self.maxSpeed, speed + Self.acceleration * dt)
    if isNitroActive {
      nitro = max(0, nitro - Self.nitroDrain * dt)
    } else if !isBoosting {
      // The gauge only refills once the pedal is released — holding it on
      // an empty tank never sputters the boost back on.
      nitro = min(1, nitro + Self.nitroRegen * dt)
    }
    let groundSpeed = effectiveSpeed
    distanceMeters += groundSpeed * dt / Self.unitsPerMeter

    // Lane-change glide: constant lateral speed toward the target center,
    // snapping on arrival so drift never accumulates.
    let target = Self.laneCenter(targetLane)
    if playerX != target {
      let delta = Self.laneChangeSpeed * dt
      if abs(target - playerX) <= delta {
        playerX = target
      } else {
        playerX += target > playerX ? delta : -delta
      }
    }

    // Traffic scrolls by the closing speed; cars behind the field are
    // recycled once fully off-screen.
    for index in traffic.indices {
      traffic[index].y += (groundSpeed - traffic[index].speed) * dt
      if !traffic[index].passed, traffic[index].y > Self.playerY + Self.carLength {
        traffic[index].passed = true
        overtakes += 1
        events.append(.overtake(total: overtakes))
      }
    }
    traffic.removeAll { $0.y >= Self.fieldHeight + Self.carLength * 2 }

    // Spawn cadence ramps with elapsed time.
    spawnCooldown -= dt
    if spawnCooldown <= 0 {
      spawnCooldown = max(
        Self.spawnIntervalMin,
        Self.spawnIntervalStart - elapsed * Self.spawnIntervalRamp
      )
      trySpawnCar()
    }

    // Crash check: forgiving rectangle overlap against every live car.
    let crashed = traffic.contains { car in
      abs(car.x - playerX) < Self.hitWidth
        && Self.playerY + Self.hitLengthPad < car.y + Self.carLength
        && Self.playerY + Self.carLength - Self.hitLengthPad > car.y
    }
    if crashed {
      isOver = true
      events.append(.crash(score: score))
    }

    return events
  }

  /// Roll a spawn (RNG order: lane, speed, kind — matching the web engine).
  /// The spawn is dropped when the rolled lane is too crowded or when it
  /// would close the last free lane in the entry window.
  private func trySpawnCar() {
    let lane = Int(random() * Double(Self.laneCount))
    let speed = Self.trafficMinSpeed + random() * (Self.trafficMaxSpeed - Self.trafficMinSpeed)
    let kind = Int(random() * Double(Self.trafficKindCount))

    let tooClose = traffic.contains { $0.lane == lane && $0.y < Self.spawnGap - Self.carLength }
    if tooClose {
      return
    }
    var blockedLanes = Set(traffic.filter { $0.y < Self.entryWindow }.map(\.lane))
    blockedLanes.insert(lane)
    if blockedLanes.count >= Self.laneCount {
      return
    }

    traffic.append(RaceTrafficCar(
      lane: lane,
      x: Self.laneCenter(lane),
      y: -Self.carLength,
      speed: speed,
      kind: kind,
      passed: false
    ))
  }

  // MARK: - Test hooks

  /// Test hook: empty the road to build exact scenarios.
  func clearTraffic() {
    traffic = []
  }

  /// Test hook: place a car at an exact spot.
  func addCar(lane: Int, y: Double, speed: Double = RaceEngine.trafficMinSpeed, kind: Int = 0) {
    traffic.append(RaceTrafficCar(
      lane: lane, x: Self.laneCenter(lane), y: y, speed: speed, kind: kind, passed: false
    ))
  }

  /// Test hook: snap the player into a lane with no glide.
  func snapToLane(_ lane: Int) {
    targetLane = lane
    playerX = Self.laneCenter(lane)
  }

  /// Test hook: roll one spawn attempt without advancing the world.
  func forceSpawnRoll() {
    trySpawnCar()
  }
}

File: android/app/src/main/kotlin/com/baseapp/android/view/games/race/RaceEngine.kt
package com.baseapp.android.view.games.race

import kotlin.math.abs
import kotlin.math.min
import kotlin.random.Random

/**
 * Pure Kotlin port of the web RaceBot simulation
 * (`web/app/src/View/Games/Race/engine.ts`) so the exact same rules run on
 * every platform and can be unit-tested on the JVM. No Android or Compose
 * imports here — rendering and input live in `RaceGameView`.
 *
 * Steering is tap-based ([steerLeft]/[steerRight] move the target lane);
 * nitro arrives as a held-input flag ([isBoosting]) — the native twin of a
 * held key.
 *
 * All coordinates are in field units (420x720, y growing downward); the view
 * scales the field to the device while preserving the aspect ratio. Time is
 * the engine's own [elapsed] clock, so identical steps always produce
 * identical states.
 *
 * Randomness (spawn lane, traffic speed, paint job) goes through an injected
 * [Random] so tests can pin every spawn decision. The RNG call order in
 * `trySpawnCar` (lane, speed, kind) matches the web engine exactly.
 */
class RaceEngine(private val random: Random = Random.Default) {
    companion object {
        // Field geometry and tuning — must stay byte-for-byte in sync with
        // the web constants in engine.ts.
        const val FIELD_WIDTH = 420f
        const val FIELD_HEIGHT = 720f
        const val LANE_COUNT = 3

        /** Asphalt inset: shoulders on both sides of the three 110-unit lanes. */
        const val ROAD_LEFT = 45f
        const val ROAD_WIDTH = 330f
        const val LANE_WIDTH = ROAD_WIDTH / LANE_COUNT
        const val CAR_WIDTH = 62f
        const val CAR_LENGTH = 108f

        /** Top edge of the player car; fixed — the world scrolls, the car does not. */
        const val PLAYER_Y = 560f

        /** Cruise speed at the green light (field units/s). */
        const val BASE_SPEED = 260f

        /** Cruise ceiling; nitro can push past it. */
        const val MAX_SPEED = 640f

        /** Cruise speed gained per second of driving — the difficulty ramp. */
        const val ACCELERATION = 9f

        /** Sideways glide speed during a lane change (units/s). */
        const val LANE_CHANGE_SPEED = 520f

        const val NITRO_MULTIPLIER = 1.6f

        /** Gauge (0-1) drained per second while boosting. */
        const val NITRO_DRAIN = 0.35f

        /** Gauge regenerated per second while cruising. */
        const val NITRO_REGEN = 0.12f

        /** Traffic cruises at a fixed speed rolled at spawn (units/s). */
        const val TRAFFIC_MIN_SPEED = 110f
        const val TRAFFIC_MAX_SPEED = 215f

        /** Distinct traffic paint jobs; the roll only picks the palette index. */
        const val TRAFFIC_KIND_COUNT = 4

        /** Seconds between spawn attempts at the start of a run… */
        const val SPAWN_INTERVAL_START = 1.35f

        /** …and the floor it ramps down to. */
        const val SPAWN_INTERVAL_MIN = 0.55f

        /** How much the spawn interval shrinks per second of driving. */
        const val SPAWN_INTERVAL_RAMP = 0.008f

        /**
         * A spawn is skipped if it would leave no free lane in the top
         * [ENTRY_WINDOW] units of the field — there must always be a way
         * through.
         */
        const val ENTRY_WINDOW = 260f

        /** Minimum bumper-to-bumper gap within a lane at spawn time. */
        const val SPAWN_GAP = CAR_LENGTH * 1.6f

        /** Score per car passed. */
        const val OVERTAKE_BONUS = 50

        /** Field units per scored meter (≈ 43 m/s at base cruise). */
        const val UNITS_PER_METER = 6f

        /** Collision paddings: slightly forgiving, like the web. */
        const val HIT_WIDTH = CAR_WIDTH * 0.82f
        const val HIT_LENGTH_PAD = 6f

        fun laneCenter(lane: Int): Float = ROAD_LEFT + LANE_WIDTH * (lane + 0.5f)
    }

    /**
     * One traffic car. `x` is the center in field units, `y` the top edge;
     * `kind` picks the paint-job palette entry.
     */
    class TrafficCar(
        /** Lane index 0 (left) to 2 (right). */
        val lane: Int,
        val x: Float,
        var y: Float,
        /** Own cruise speed (units/s); closing speed is player minus this. */
        val speed: Float,
        /** Paint-job palette index (0..TRAFFIC_KIND_COUNT-1). */
        val kind: Int,
        /** Whether the overtake bonus for this car has been paid. */
        var passed: Boolean = false,
    )

    /**
     * Discrete things that happened during one [step] — the native twin of
     * the web game's HUD/crash callbacks.
     */
    data class StepResult(
        /** Total overtakes after each car passed this step (usually 0-1). */
        val overtakeTotals: List<Int> = emptyList(),
        val isGameOver: Boolean = false,
    )

    /** Held input: nitro pedal down. */
    var isBoosting = false

    val traffic = mutableListOf<TrafficCar>()

    /** Lane the player is gliding toward. */
    var targetLane = 1
        private set

    /** Player center x; glides toward the target lane's center. */
    var playerX = laneCenter(1)
        private set

    /** Current cruise speed (before the nitro multiplier). */
    var speed = BASE_SPEED
        private set

    /** Nitro gauge, 0-1. */
    var nitro = 1f
        private set

    var distanceMeters = 0f
        private set

    var overtakes = 0
        private set

    var isOver = false
        private set

    /** Engine clock in seconds; drives the difficulty ramp. */
    var elapsed = 0f
        private set

    private var spawnCooldown = SPAWN_INTERVAL_START

    val score: Int
        get() = distanceMeters.toInt() + overtakes * OVERTAKE_BONUS

    /** True while the nitro flame should render and the multiplier applies. */
    val isNitroActive: Boolean
        get() = isBoosting && nitro > 0f && !isOver

    /** Ground speed including nitro (what the speedometer shows). */
    val effectiveSpeed: Float
        get() = speed * (if (isNitroActive) NITRO_MULTIPLIER else 1f)

    /** Full reset (the web "Restart Race" path). */
    fun newGame() {
        traffic.clear()
        targetLane = 1
        playerX = laneCenter(1)
        speed = BASE_SPEED
        nitro = 1f
        distanceMeters = 0f
        overtakes = 0
        isOver = false
        elapsed = 0f
        spawnCooldown = SPAWN_INTERVAL_START
        isBoosting = false
    }

    fun steerLeft() {
        if (!isOver && targetLane > 0) {
            targetLane -= 1
        }
    }

    fun steerRight() {
        if (!isOver && targetLane < LANE_COUNT - 1) {
            targetLane += 1
        }
    }

    /**
     * Advance the simulation. Mirrors the web `step` frame-for-frame: ramp
     * the cruise speed, drain/refill nitro, glide the lane change, scroll
     * traffic by the closing speed, pay overtakes, spawn on the cadence,
     * and check for a crash.
     */
    fun step(dt: Float): StepResult {
        if (isOver) {
            return StepResult()
        }
        val overtakeTotals = mutableListOf<Int>()
        elapsed += dt

        // Cruise ramp and nitro gauge.
        speed = min(MAX_SPEED, speed + ACCELERATION * dt)
        nitro = when {
            isNitroActive -> (nitro - NITRO_DRAIN * dt).coerceAtLeast(0f)
            // The gauge only refills once the pedal is released — holding
            // it on an empty tank never sputters the boost back on.
            !isBoosting -> (nitro + NITRO_REGEN * dt).coerceAtMost(1f)
            else -> nitro
        }
        val groundSpeed = effectiveSpeed
        distanceMeters += groundSpeed * dt / UNITS_PER_METER

        // Lane-change glide: constant lateral speed toward the target
        // center, snapping on arrival so drift never accumulates.
        val target = laneCenter(targetLane)
        if (playerX != target) {
            val delta = LANE_CHANGE_SPEED * dt
            playerX = if (abs(target - playerX) <= delta) {
                target
            } else {
                playerX + (if (target > playerX) delta else -delta)
            }
        }

        // Traffic scrolls by the closing speed; cars behind the field are
        // recycled once fully off-screen.
        for (car in traffic) {
            car.y += (groundSpeed - car.speed) * dt
            if (!car.passed && car.y > PLAYER_Y + CAR_LENGTH) {
                car.passed = true
                overtakes += 1
                overtakeTotals.add(overtakes)
            }
        }
        traffic.removeAll { it.y >= FIELD_HEIGHT + CAR_LENGTH * 2 }

        // Spawn cadence ramps with elapsed time.
        spawnCooldown -= dt
        if (spawnCooldown <= 0f) {
            spawnCooldown = (SPAWN_INTERVAL_START - elapsed * SPAWN_INTERVAL_RAMP)
                .coerceAtLeast(SPAWN_INTERVAL_MIN)
            trySpawnCar()
        }

        // Crash check: forgiving rectangle overlap against every live car.
        val crashed = traffic.any { car ->
            abs(car.x - playerX) < HIT_WIDTH &&
                PLAYER_Y + HIT_LENGTH_PAD < car.y + CAR_LENGTH &&
                PLAYER_Y + CAR_LENGTH - HIT_LENGTH_PAD > car.y
        }
        if (crashed) {
            isOver = true
        }

        return StepResult(overtakeTotals = overtakeTotals, isGameOver = crashed)
    }

    /**
     * Roll a spawn (RNG order: lane, speed, kind — matching the web engine).
     * The spawn is dropped when the rolled lane is too crowded or when it
     * would close the last free lane in the entry window.
     */
    private fun trySpawnCar() {
        val lane = (random.nextFloat() * LANE_COUNT).toInt()
        val speed = TRAFFIC_MIN_SPEED + random.nextFloat() * (TRAFFIC_MAX_SPEED - TRAFFIC_MIN_SPEED)
        val kind = (random.nextFloat() * TRAFFIC_KIND_COUNT).toInt()

        val tooClose = traffic.any { it.lane == lane && it.y < SPAWN_GAP - CAR_LENGTH }
        if (tooClose) {
            return
        }
        val blockedLanes = traffic.filter { it.y < ENTRY_WINDOW }.map { it.lane }.toMutableSet()
        blockedLanes.add(lane)
        if (blockedLanes.size >= LANE_COUNT) {
            return
        }

        traffic.add(
            TrafficCar(lane = lane, x = laneCenter(lane), y = -CAR_LENGTH, speed = speed, kind = kind),
        )
    }

    // Test hooks -----------------------------------------------------------

    /** Test hook: empty the road to build exact scenarios. */
    fun clearTraffic() {
        traffic.clear()
    }

    /** Test hook: place a car at an exact spot. */
    fun addCar(lane: Int, y: Float, speed: Float = TRAFFIC_MIN_SPEED, kind: Int = 0) {
        traffic.add(TrafficCar(lane = lane, x = laneCenter(lane), y = y, speed = speed, kind = kind))
    }

    /** Test hook: snap the player into a lane with no glide. */
    fun snapToLane(lane: Int) {
        targetLane = lane
        playerX = laneCenter(lane)
    }

    /** Test hook: roll one spawn attempt without advancing the world. */
    fun forceSpawnRoll() {
        trySpawnCar()
    }
}
